File: flutter_prayer_times/android/app/src/main/java/com/example/flutter_prayer_times/rest/RestServerFactory.java
package com.example.flutter_prayer_times.rest;

import android.content.Context;
import android.widget.Toast;

import org.json.simple.JSONArray;

import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

public class RestServerFactory {

    Context ctxt = null;

    RestServerFactory(Context ctxt){
        this.ctxt = ctxt;
    }

    private static final String BASE_URL = "https://prayer-times.vsyou.app";


    public static RestServicePrayerTimes createRestService() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(JacksonConverterFactory.create())
                .build();

        return retrofit.create(RestServicePrayerTimes.class);
    }

    public static void getDataFromServer(Context ctxt) {
        RestServicePrayerTimes restServicePrayerTimes = createRestService();

        Call<List<Map<String, List<String>>>> call = restServicePrayerTimes.getAll();

        call.enqueue(new Callback<List<Map<String, List<String>>>>() {

            @Override
            public void onResponse(
                    Call<List<Map<String, List<String>>>> call,
                    Response<List<Map<String, List<String>>>> response
            ) {
                if (!response.isSuccessful()) {
                    return;
                }

                List<Map<String, List<String>>> resBody = response.body();

                System.out.println("List of API" + resBody);

                JSONArray listOfPrayers = new JSONArray();
                for (int i = 0; i < resBody.toArray().length; i++) {
                    listOfPrayers.add(resBody.toArray()[i]);
                }

                try (FileWriter file = new FileWriter("/data/user/0/com.example.flutter_prayer_times/app_flutter/serverDataFile.json")) {
                    file.write(listOfPrayers.toJSONString());
                    System.out.println("Successfully Copied JSON Object to File...");
                    Toast.makeText(ctxt, "Prayer Times are updated", Toast.LENGTH_SHORT).show();
                    System.out.println("JSON Object: " + listOfPrayers);
                } catch (IOException e) {
                    e.printStackTrace();
                }
                return;
            }

            @Override
            public void onFailure(
                    Call<List<Map<String, List<String>>>> call,
                    Throwable t) {
                System.out.println("TestTest");
                System.out.println(t);
            }
        });
    }

}

File: flutter_prayer_times/android/app/src/main/java/com/example/flutter_prayer_times/rest/RestServerFactory.kt
package com.example.flutter_prayer_times.rest

import android.annotation.SuppressLint
import android.content.Context
import android.content.ContextWrapper
import android.os.Build
import android.support.annotation.RequiresApi
import com.example.flutter_prayer_times.AppSettings.AppSettings
import com.example.flutter_prayer_times.JsonFilesServices
import com.google.gson.JsonParser
import retrofit2.Call
import retrofit2.Callback
import java.io.BufferedReader
import java.io.File
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.*
import retrofit2.Response as Response1


class RestServerFactory internal constructor() {
    companion object {
        val restServicePrayerTimes = RestServicePrayerTimes.create()
        fun getDataFromServer(ctxt: Context, force: Boolean = false) {
            val appSettings = AppSettings(ctxt)
            val call = restServicePrayerTimes.all(
                    lat = appSettings.lat.toString(),
                    lng = appSettings.lng.toString(),
                    language = appSettings.getDataFromAppSettingsFile().languages!!.selected)

            call.enqueue(object : Callback<Map<String, List<Map<String, List<String>>>>> {
                @SuppressLint("SimpleDateFormat")
                override fun onResponse(
                        call: Call<Map<String, List<Map<String, List<String>>>>>,
                        response: Response1<Map<String, List<Map<String, List<String>>>>>
                ) {
                    if (!response.isSuccessful)
                        return

                    val resBody = response.body()
                    val fileName = "/prayerTimesJsonDateInMonth.json"
                    val contextWrapper = ContextWrapper(ctxt)
                    val filePath = contextWrapper.getDir("flutter", 0).path

                    var checkIfPrayerDateIsAvailable = true
                    if (File(filePath + fileName).exists()) {
                        val bufferedReader: BufferedReader = File(filePath + fileName).bufferedReader()
                        // Read the text from bufferReader and store in String variable
                        val yourJson = bufferedReader.use { it.readText() }
                        val parser = JsonParser()
                        val element = parser.parse(yourJson)
                        val obj = element!!.getAsJsonObject()
                        val todayDate = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            val formatters = DateTimeFormatter.ofPattern("yyyy-MM-dd")
                            LocalDate.now().format(formatters)
                        } else{
                            val formatters = SimpleDateFormat("yyyy-MM-dd")
                            formatters.format(Date())
                        }
                        checkIfPrayerDateIsAvailable = todayDate in obj.keySet()
                    }

                    val isWrited = JsonFilesServices.writeToJsonFile(
                            ctxt = ctxt,
                            fileName = fileName,
                            jsonObject = resBody,
                            checkQuery = checkIfPrayerDateIsAvailable,
                            checkMessage = "data in $fileName are up to date",
                            force = force)

                    if (isWrited)
                        JsonFilesServices.getTodayDatesFromJsonFile(ctxt)
                    return
                }

                override fun onFailure(
                        call: Call<Map<String, List<Map<String, List<String>>>>>,
                        t: Throwable) {
                    println(t)
                }
            })
        }
    }
}
